### PR TITLE
sql: Persist DB names to view descriptors with new Format flag

### PR DESCRIPTION
**Pull request by a-robinson**
We don't have to go this way, but as explained in the comment thread on https://github.com/cockroachdb/cockroach/issues/9921#issuecomment-254009800, this was enough easier than trying to adapt Walk that I thought it was worth putting together as a proof of concept, at the very least.

@knz

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/10009)

### CODE REVIEW

**Comment by knz** [comment]
Wow, color me amazed. This is a rather nifty/elegant trick. Congrats on getting this to work with such a minimal-but-yet-rather-smart change.

:lgtm:

---

Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/create.go, line 934 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10009#-KUD8Da0BFk_qBOV2YsU:-KUD8Da0BFk_qBOV2YsV:b9hj7kv) ([raw file](https://github.com/cockroachdb/cockroach/blob/63d49ca328a9b07ffcb96d1b555a2784cd605d15/pkg/sql/create.go#L934)):*
> ```Go
> 			tn, err := n.p.QualifyWithDatabase(t)
> 			if err != nil {
> 				log.Warningf(n.p.ctx(), "Failed to qualify table name %q with database name: %v", t, err)
> ```

for consistency, use small letter "f" at the start of the message.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10009#-:-KUD84HeDtodQXagx0WM:b1s6ilv)*

**Comment by knz** [comment]
@nvanbenschoten check this out, what do you think?

**Comment by knz** [comment]
Fixes #9921.

File: pkg/sql/create.go [hunk 1]
```diff
@@ -925,7 +926,20 @@ func (n *createViewNode) makeViewTableDesc(
 	n.resolveViewDependencies(&desc, affected)
 
 	var buf bytes.Buffer
-	p.AsSource.Format(&buf, parser.FmtSimple)
+	var fmtErr error
+	p.AsSource.Format(&buf, parser.FmtNormalizeTableNames(
+		func(t *parser.NormalizableTableName) *parser.TableName {
+			tn, err := n.p.QualifyWithDatabase(t)
+			if err != nil {
+				log.Warningf(n.p.ctx(), "Failed to qualify table name %q with database name: %v", t, err)
```
Comment (tschottdorf) [inline pkg/sql/create.go]: s/F/f/

**Comment by nvanbenschoten** [comment]
This seems pretty neat, and as long as @knz doesn't feel that it is abusing the parser.Format infrastructure, I'm fully onboard!

Also, with respect to the issue this fixes -- if we are going to close that, then we should open new issues for some of the other points @a-robinson brought up (eg. the star selector complication).

---

Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/create.go, line 934 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10009#-KUFJVDTxoX6mot4_SqE:-KUFJVDTxoX6mot4_SqF:b4h0nyp) ([raw file](https://github.com/cockroachdb/cockroach/blob/63d49ca328a9b07ffcb96d1b555a2784cd605d15/pkg/sql/create.go#L934)):*
> ```Go
> 			tn, err := n.p.QualifyWithDatabase(t)
> 			if err != nil {
> 				log.Warningf(n.p.ctx(), "Failed to qualify table name %q with database name: %v", t, err)
> ```

I'm appalled no one has caught this yet: `s/Failed/failed`

---

*[pkg/sql/parser/format.go, line 27 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10009#-KUFIXblTDM5XErGSdC5:-KUFIXblTDM5XErGSdC6:b-liaq0u) ([raw file](https://github.com/cockroachdb/cockroach/blob/63d49ca328a9b07ffcb96d1b555a2784cd605d15/pkg/sql/parser/format.go#L27)):*
> ```Go
> 	showTypes           bool
> 	showTableAliases    bool
> 	tableNameNormalizer func(*NormalizableTableName) *TableName
> ```

It should be documented somewhere (assuming the pattern you have below is correct and you didn't mean to panic inside the else block) that the contract for this field is to call the function if it exists but simply ignore the result if it returns nil.

---

*[pkg/sql/parser/table_name.go, line 45 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10009#-KUFIFinOxjwDG0gn7_3:-KUFIFioEyakJgP3sQOJ:b-e43jmp) ([raw file](https://github.com/cockroachdb/cockroach/blob/63d49ca328a9b07ffcb96d1b555a2784cd605d15/pkg/sql/parser/table_name.go#L45)):*
> ```Go
> 		if tn := f.tableNameNormalizer(&nt); tn != nil {
> 			tnr = tn
> 		} else {
> ```

There's no need for this `else` block

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10009#-:-KUFJpCl5jLTvmsvKf_8:b-id9qlr)*

**Comment by a-robinson** [comment]
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/create.go, line 934 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10009#-KUD8Da0BFk_qBOV2YsU:-KUHh3ilvD4RIWbzrATL:b-rswckh) ([raw file](https://github.com/cockroachdb/cockroach/blob/63d49ca328a9b07ffcb96d1b555a2784cd605d15/pkg/sql/create.go#L934)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> for consistency, use small letter "f" at the start of the message.
</details>
I probably should have noticed by now from having spent so much time reading logs, but I'm a little surprised that we don't follow the standard golang style recommendation on this (errors start lower case, logs start upper case - https://github.com/golang/go/wiki/CodeReviewComments#error-strings)

---

*[pkg/sql/create.go, line 934 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10009#-KUFJVDTxoX6mot4_SqE:-KUHhMNu7wIwA0k8RcHb:bxp9gxv) ([raw file](https://github.com/cockroachdb/cockroach/blob/63d49ca328a9b07ffcb96d1b555a2784cd605d15/pkg/sql/create.go#L934)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> I'm appalled no one has caught this yet: `s/Failed/failed`
</details>
Heh, thanks

---

*[pkg/sql/parser/format.go, line 27 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10009#-KUFIXblTDM5XErGSdC5:-KUHiA4OH6xCwDXLyRJR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/63d49ca328a9b07ffcb96d1b555a2784cd605d15/pkg/sql/parser/format.go#L27)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> It should be documented somewhere (assuming the pattern you have below is correct and you didn't mean to panic inside the else block) that the contract for this field is to call the function if it exists but simply ignore the result if it returns nil.
</details>
Done.

---

*[pkg/sql/parser/table_name.go, line 45 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10009#-KUFIFinOxjwDG0gn7_3:-KUHhjP_W9eTvqVeD0w8:b-6jncsy) ([raw file](https://github.com/cockroachdb/cockroach/blob/63d49ca328a9b07ffcb96d1b555a2784cd605d15/pkg/sql/parser/table_name.go#L45)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> There's no need for this `else` block
</details>
Thanks for catching that! I had some debug logging in there and didn't clean it up properly.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10009)*

**Pull request merged**